File: solutions/2026-03-27/cpp/2026-03-27-matrix-similarity-after-cyclic-shifts.cpp
```cpp
/*
Problem Summary:
Given a matrix and an integer k, determine if the matrix remains the same after performing k cyclic shifts on its rows.
Even-indexed rows shift left, and odd-indexed rows shift right.

Link: https://leetcode.com/problems/matrix-similarity-after-cyclic-shifts/

Approach Explanation:
The core idea is to realize that after a certain number of shifts, the rows will return to their original positions.
For an even-indexed row of length 'n', a left shift by 'n' positions brings it back to its original state.
For an odd-indexed row of length 'n', a right shift by 'n' positions brings it back to its original state.
Therefore, we only need to consider k modulo n for each row's length. If k is a multiple of n, the row will be the same.
If k modulo n is 0 for all rows, then the entire matrix will be the same.
We iterate through each row of the matrix. For even-indexed rows, we check if k % n == 0. For odd-indexed rows, we also check if k % n == 0.
If this condition is not met for any row, the matrix will change, and we return false. If the condition holds for all rows, we return true.

Time Complexity Analysis:
The time complexity is O(m), where 'm' is the number of rows in the matrix. This is because we iterate through each row once to check the condition k % n. The length of the rows 'n' does not affect the overall time complexity as we are only performing a modulo operation.

Space Complexity Analysis:
The space complexity is O(1) because we are only using a few variables to store the number of rows, columns, and the shift count. We do not create any auxiliary data structures that grow with the input size.
*/
#include <vector>
#include <numeric>

class Solution {
public:
    bool differByOne(std::vector<std::vector<int>>& mat, int k) {
        int m = mat.size(); // Number of rows
        if (m == 0) {
            return true; // Empty matrix is considered similar
        }
        int n = mat[0].size(); // Number of columns

        // If k is 0, no shifts occur, so the matrix is always the same.
        if (k == 0) {
            return true;
        }

        // Iterate through each row of the matrix
        for (int i = 0; i < m; ++i) {
            // Check if the row index is even or odd
            if (i % 2 == 0) {
                // Even-indexed rows are shifted left.
                // A left shift by 'n' positions returns the row to its original state.
                // So, if k % n is not 0, the row will be different from its original state.
                if (k % n != 0) {
                    return false; // Matrix will differ
                }
            } else {
                // Odd-indexed rows are shifted right.
                // A right shift by 'n' positions returns the row to its original state.
                // So, if k % n is not 0, the row will be different from its original state.
                if (k % n != 0) {
                    return false; // Matrix will differ
                }
            }
        }

        // If we have checked all rows and none of them will be different,
        // then the entire matrix remains the same.
        return true;
    }
};
```

File: solutions/2026-03-27/javascript/2026-03-27-matrix-similarity-after-cyclic-shifts.js
/**
 * @param {number[][]} mat
 * @param {number} k
 * @return {boolean}
 */
// Problem Summary: Check if a matrix returns to its original state after k cyclic shifts on even and odd rows.
// Link: https://leetcode.com/problems/matrix-similarity-after-cyclic-shifts/
// Approach:
// The core idea is to realize that for the matrix to be identical after k shifts, each row must return to its original state after k shifts.
// For even-indexed rows, a left cyclic shift by 1 moves the first element to the end. After `mat[0].length` shifts, the row returns to its original state.
// For odd-indexed rows, a right cyclic shift by 1 moves the last element to the beginning. After `mat[1].length` shifts, the row returns to its original state.
// Therefore, we only need to check if `k` is a multiple of the row length for each row. If `k % row.length === 0`, then that row will return to its original state.
// We iterate through each row. If it's an even-indexed row, we check if `k % mat[i].length === 0`. If it's an odd-indexed row, we also check if `k % mat[i].length === 0`.
// If for any row, `k % mat[i].length !== 0`, then that row will not return to its original state, and thus the entire matrix will not be similar. We can immediately return `false`.
// If we iterate through all rows and `k` is a multiple of every row's length, it means all rows will return to their original positions, and we return `true`.
//
// Time Complexity: O(m), where m is the number of rows. We iterate through each row once. The modulo operation takes constant time.
// Space Complexity: O(1). We are not using any extra space that depends on the input size.
const matrix1758 = (mat, k) => {
    const m = mat.length;
    // Iterate through each row of the matrix
    for (let i = 0; i < m; i++) {
        // Check if the number of shifts k is a multiple of the row's length.
        // If k is a multiple of the row length, then after k shifts, the row will return to its original state.
        // This is because a cyclic shift by `row.length` positions brings the row back to its original configuration.
        if (k % mat[i].length !== 0) {
            // If for any row, k is not a multiple of its length,
            // that row will not be in its original state after k shifts.
            // Therefore, the entire matrix will not be similar to the original.
            return false;
        }
    }
    // If k is a multiple of every row's length, then all rows will return to their original states.
    // Thus, the matrix will be identical to the original.
    return true;
};
```